### PR TITLE
[Home Page Picker] Connect Choose and Skip Actions in site creation

### PR DESCRIPTION
**Pull request by chipsnyder**
**Fixes** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2721

**Related PR**
- `WordPressKit-iOS`: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/299

📓 This handles the Blog flag as well and skip button. There are two backend changes that are pending that I'm waiting on to make sure this is fully testable.
## To test:
<details>
<summary><strong>To disable or enable the development version of Home Page Picker</strong></summary>

- Open the app from the build that allows FeatureFlags such as a PR build or a local development build
    - By default, the modal layout picker will be disabled.
    - From the site page:
        - Click on your Gravatar.
        - Click on App Settings.
        - Click on Debug.
        - Toggle "Site Creation: Home Page Picker" to enable or disable the picker
        - <kdb><a href="https://user-images.githubusercontent.com/3384451/97343011-5f032980-185d-11eb-9030-06c17f732fec.gif"></a></kdb>

</details>

When enabled, the first step of the Site Creation flow will be replaced with the Home Page Picker flow.

### Create a Site with Site Design
1. Navigate to the Create WordPress.com site
    - My Sites > ➕ > Create WordPress.com site
1. Select a design
1. Tap on "Choose"
1. Search for and Choose a domain
    - **Expect** to see `.wordpress.com` subdomains only unless you selected an option from the test `Create a Blog`
1. Tap "Create Site"

**Expect** Your site to be created

### Create a Site by skipping site design
1. Navigate to the Create WordPress.com site
    - My Sites > ➕ > Create WordPress.com site
1. Select "Skip"
1. Search for and Choose a domain
1. Tap "Create Site"
**Expect** Your site to be created

### Create a Blog
1. Navigate to the Create WordPress.com site
    - My Sites > ➕ > Create WordPress.com site
1. Select a design that has been flagged as a blog
    <details>
    <summary>Examples</summary>
    <kdb><a href="https://user-images.githubusercontent.com/3384451/98397556-87451200-202d-11eb-92f6-c42e86a1a7d2.png"></a></kdb>
    </details>
1. Tap on "Choose"
1. Search for and Choose a domain
    - **Expect** to see `.wordpress.com` and `.blog` subdomains.
1. Tap "Create Site"
**Expect** Your site to be created

## Validate existing flow:

To make sure nothing is broken in the existing flows (in case we want to preserve it for A/B test scenarios). Disable the Home page picker feature. 

### Create a Site
1. Navigate to the Create WordPress.com site
    - My Sites > ➕ > Create WordPress.com site
1. Select a segment other than Blog
1. Search for and Choose a domain
    - **Expect** to see `.wordpress.com` subdomains only.
1. Tap "Create Site"
**Expect** Your site to be created

### Create a Blog
1. Navigate to the Create WordPress.com site
    - My Sites > ➕ > Create WordPress.com site
1. Select the Blog segment
1. Search for and Choose a domain
    - **Expect** to see `.wordpress.com` and `.blog` subdomains.

## Screenshots:

Site Creation | Site for "Skip"
--- | ---
<kdb><a href="https://user-images.githubusercontent.com/3384451/98302254-113b9f00-1f8a-11eb-92f6-b213c3f971d8.gif"></a></kdb> | <kdb><a href="https://user-images.githubusercontent.com/3384451/98852584-df14bc00-2425-11eb-8810-c2e866117414.png"></a></kdb>

## Additional Context: 

This is a continuation of the work done in the [Home Page Picker Project](https://github.com/wordpress-mobile/gutenberg-mobile/labels/Project%3A%20Home%20Page%20Picker). 

## PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by chipsnyder** [comment]
@kyleaparker @iamthomasbishop @SylvesterWilmott I'm still working on a few more things as part of this PR, but if you'd like to start taking a look at the Home Page picker, this PR should be a good entry point to start providing any feedback you have.

**Comment by chipsnyder** [comment]
Hey @antonis and @mkevins I have this pretty much ready but there are a few test cases here that we can't fully validate until some other work is done. But if you'd like to take an initial look then this is basically good to go.

**Comment by antonis** [comment]
Great work @chipsnyder 👍 
The app works as expected and the code LGTM 🎉 

I noticed two minor UI issues that are not directly related with this PR:

<details>
<summary>1. Landscape mode breaks the UI on iPhone</summary>
<image width="680" src="https://user-images.githubusercontent.com/304044/98560017-fa5dbb00-22af-11eb-9eaa-e2ffdd3b7c06.gif"/>
</details>

<details>
<summary>2. On iPad I'm not sure if the choose toolbar button should fit the screen width</summary>
<image width="680" src="https://user-images.githubusercontent.com/304044/98559920-ddc18300-22af-11eb-82ff-fa2fba7d14fd.png"/>
</details>

**Comment by kyleaparker** [comment]
@chipsnyder this is looking really good! For some reason I'm not seeing the Skip/Back buttons in the top bar: 

98590375-9a105d00-2333-11eb-9fdb-2e2667ae36dc PNG (1125×2436) 2020-11-10 10-05-06

There may be another setting I need to switch somewhere because I still have the blue navigation bar on the choose domain screen, and it seems to be white in your examples. 

Some other early feedback: 
- The default design when skipped should be: https://defaultstartermobile.wordpress.com/
- I'm wondering if we should allow people to skip the choose a domain step as well. I also mentioned this in the [Android PR](https://github.com/wordpress-mobile/WordPress-Android/pull/13316). 
- There is a divider on the choose a design screen, underneath the subtext/above the screenshots that should be removed

**Comment by iamthomasbishop** [comment]
@chipsnyder Stoked to see this in action! I think this aligns well with designs that @SylvesterWilmott worked on, but I'm curious now that I see it in action — would it make sense to have this in a modal sheet presentation like the other layout picker?

> There may be another setting I need to switch somewhere because I still have the blue navigation bar on the choose domain screen, and it seems to be white in your examples.

I was also seeing some weirdness w/ the blue navbar as @kyleaparker mentioned, but I think it might be because the icon and text label are white (in both modes) — I am seeing them as expected in dark mode.

> There is a divider on the choose a design screen, underneath the subtext/above the screenshots that should be removed

I agree w/ @kyleaparker here. We do have that border on other layout picker, but it feels more natural in that context (with the category buttons). Can we hide it until the nav bar becomes fixed?

This also made me realize the styling of the nav bar on the domain step is slightly different (probably using the default nav bar style we have elsewhere) — which begs the question, should we extend the styling of the nav bar on the previous step (homepage selection) to this for consistency? (eg. regular system material, bordering style)

**Comment by chipsnyder** [comment]
@antonis

> 1. Landscape mode breaks the UI on iPhone

Nice catch there! This is fixed now :)

> 2. On iPad I'm not sure if the choose toolbar button should fit the screen width

This does become a bit much. Maybe setting a maximum width on this button to match the width of cells would be better. 

@iamthomasbishop suggested using the same modal presentation style that MLP uses, which might help with this. @iamthomasbishop can you check out the latest build on an iPad and let us know what you think?

---

@kyleaparker 
> The default design when skipped should be: https://defaultstartermobile.wordpress.com/

Thanks for confirming this! @tug and @enej are working on a patch on the backend to ensure this flow uses the correct design when we skip. I'll drop another note here or in an updated PR once that's ready. 

> I'm wondering if we should allow people to skip the choose a domain step as well. I also mentioned this in the Android PR.

It looks like this is an option if you create a site on web. When I created a site we ended up with a site that was my username +  random number. So this might be an option but I would like to do some more exploration before adding it to this iteration. Maybe we can add it as an iteration 2 item.

>There is a divider on the choose a design screen, underneath the subtext/above the screenshots that should be removed

Yup, that's not a big deal. I made that adjustment in the latest build.

---

@iamthomasbishop
> would it make sense to have this in a modal sheet presentation like the other layout picker?

I like that idea, although there might be some issues to consider. That presentation style would force us into supporting landscape mode as well. The choose a domain seems to work ok but not great in landscape mode, and the final confirmation looks a bit funky. I went ahead and set iPad to use the same setting as MLP, which I think looks better. Let me know though what you think about moving forward with iPhones.

<kdb><a href="https://user-images.githubusercontent.com/3384451/98733255-3fe0bd80-236e-11eb-9969-625841ceb99b.png"></a></kdb>

> I was also seeing some weirdness w/ the blue navbar as @kyleaparker mentioned, but I think it might be because the icon and text label are white (in both modes) — I am seeing them as expected in dark mode.

Yeah, this ended up being an issue with the newNavBarStyle feature, which is turned on for developers, but I think it might be off for the generated builds. Thanks for catching this @kyleaparker and @iamthomasbishop! 

> This also made me realize the styling of the nav bar on the domain step is slightly different (probably using the default nav bar style we have elsewhere) — which begs the question, should we extend the styling of the nav bar on the previous step (homepage selection) to this for consistency? (eg. regular system material, bordering style)

I haven't gotten to a few of the style changes for the Choose a Domain screen. To help make sure we had all of the necessary changes done in time, we moved those to phase 2. 

But the switch to blue in NavBar isn't ideal, so I went ahead and used the same NavBar styling for this entire flow that we're using in MLP. I'll tackle the other design changes for that screen once we get this PR wrapped up.

**Comment by chipsnyder** [comment]
@antonis @mkevins `segement_id` is now available on the backend. ~The only piece of this that remain untestable is using the key `default` in the skip flow. If that's ok with you, I'd like to mark this ready for review and address any issues that come up for skip in a later PR.~ 

I think for design, we're good to go and we'll sync with @SylvesterWilmott on @iamthomasbishop's suggestion of using the modal window. Then tackle that later if needed. 

---
**Edit:** `default` is working on the backend now

**Comment by iamthomasbishop** [comment]
> I went ahead and set iPad to use the same setting as MLP, which I think looks better. Let me know though what you think about moving forward with iPhones. 

@chipsnyder That looks great on iPad and I think it would also work well on phone. I'd still like to see if @SylvesterWilmott and @kyleaparker have any feelings one way or the other (modal sheet presentation vs. full-screen view).

> I haven't gotten to a few of the style changes for the Choose a Domain screen. To help make sure we had all of the necessary changes done in time, we moved those to phase 2.

One more note, and if you've already got it on your radar, feel free to ignore me 😄 I was going to suggest that we might want to use the same close-button style as on the other MLP (filled circle bg w/ x in the middle), esp if we go with the modal sheet presentation.

**Comment by chipsnyder** [comment]
> I was going to suggest that we might want to use the same close-button style as on the other MLP (filled circle bg w/ x in the middle), esp if we go with the modal sheet presentation.

That's a good callout. I went back and forth on what we might want there. Since the left-hand close buttons are more the ✖️ with no background. But I'm happy to make that change.

---

This is available now in the build that will finish shortly.

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/GutenbergLayoutPickerViewController.swift [hunk 1]
```diff
@@ -79,28 +79,7 @@ class GutenbergLayoutPickerViewController: CollapsableHeaderViewController {
...
-                    return UIColor.quaternarySystemFill
-                }
-            }
-        } else {
-            closeButton.tintColor = .textSubtle
-            closeButton.backgroundColor = .quaternaryBackground
-        }
-
-        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: closeButton)
-
```
Comment (mkevins) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/GutenbergLayoutPickerViewController.swift:103]: :heart: seeing this DRYed up :smile:

File: WordPress/Classes/ViewRelated/Site Creation/WebAddress/WebAddressWizardContent.swift [hunk 2]
```diff
@@ -169,10 +169,9 @@ final class WebAddressWizardContent: UIViewController {
     }
 
     private func fetchAddresses(_ searchTerm: String) {
-        // It's not ideal to let the segment ID be optional at this point, but in order to avoid overcomplicating my current
-        // task, I'll default to silencing this situation.  Since the segment ID should exist, this silencing should not
-        // really be triggered for now.
-        guard let segmentID = siteCreator.segment?.identifier else {
+        // If the segment ID isn't set (which could happen in the case of the user skipping the site design selection) we'll fall through and search for dotcom only domains.
+        guard let segmentID = siteCreator.segment?.identifier ?? siteCreator.design?.segmentID else {
+            fetchDotComAddresses(searchTerm)
```
Comment (mkevins) [inline WordPress/Classes/ViewRelated/Site Creation/WebAddress/WebAddressWizardContent.swift:174]: I wonder about the previous comment, and if there are former cases where segmentId was `nil`.. are there other paths that lead here with `nil`, that we need to have a NOOP for? (i.e. it says segmentId "should exist", but the presense of the guard suggests the possibility of an unexpected occurance :thinking: ).

**Review by mkevins — COMMENTED** [review]
Hi Chip! Nice work on this. The code looks good, just had a question about an older code comment.

I tested on all the flows described, and everything works regarding the domain search, as expected. Other than that, a couple of things I noticed:

At the end of the flow, the site appeared blank for a long time (in some cases, seemingly indefinitely).. but the internet today has been fairly unstable (lots of sites had outages, and rumors circulating of DDOS), so perhaps it can be chalked up to that :man_shrugging:, in any case, it looked like this:



While testing the "existing" flow, the domain seach worked as expected, but the text for the back button was "Create site", which I believe is a bug:

File: WordPress/Classes/ViewRelated/Site Creation/WebAddress/WebAddressWizardContent.swift (hunk 2 above)
Comment (chipsnyder) [reply WordPress/Classes/ViewRelated/Site Creation/WebAddress/WebAddressWizardContent.swift:174]: I was curious about this as well to check and see if there were any cases I should plan for. I ended up tracking this to https://github.com/wordpress-mobile/WordPress-iOS/pull/11365 which didn't have too many additional insights. Looking deeper the `segment_id` is a [required](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/638cab2d5abc0f63da0de51431d00d9354b029d4/WordPressKit/WordPressComServiceRemote%2BSiteSegments.swift#L3) parameter on the old API.

So I chalked it up to the `siteCreator` is the parent object that holds the selected segment, so this was just a fall through to handle the needed nullability of that parameter.

**Comment by chipsnyder** [comment]
> At the end of the flow, the site appeared blank for a long time (in some cases, seemingly indefinitely).. but the internet today has been fairly unstable (lots of sites had outages, and rumors circulating of DDOS), so perhaps it can be chalked up to that 🤷‍♂️, in any case, it looked like this:

Interesting. I ran through the flows again and I'm having success. Maybe it is the connection issues you described. Can you run through another test?

>While testing the "existing" flow, the domain search worked as expected, but the text for the back button was "Create site", which I believe is a bug:

On the majority of our pages, we do use the title of the previous page as the back item prompt like we do in the "existing flow." For this updated flow using the title of the main screen `Choose a Domain` ends up making the title cramped imo; I'm not sure `Create Site` would make the most sense in the new flow since it's no longer the page title of page one. We don't use `Back` in a lot of spaces but I think here fits the needs to allow it to be one of those one-off cases. Totally willing to change this up though.

iPhone 11 | iPhone SE
--- | ---
<kdb><a href="https://user-images.githubusercontent.com/3384451/98967036-fad99a00-24d9-11eb-8e8b-b36accb4691a.png"></a></kdb> | <kdb><a href="https://user-images.githubusercontent.com/3384451/98967644-b5699c80-24da-11eb-9fb7-d85253a1451e.png"></a></kdb>

**Comment by iamthomasbishop** [comment]
@chipsnyder @mkevins Regarding the label, if we're considering a one-off custom label, maybe something like `Design` or `Choose design`. `Back` would be fine as a fall-back, but a shortened version or descriptor of the previous page would be better, imo.

**Comment by chipsnyder** [comment]
Thanks @iamthomasbishop, I'll go with `Choose design` we're using `Back` on MLP as well so be consistent there I'll update that to `Choose layout`.

**Comment by mkevins** [comment]
@chipsnyder and @iamthomasbishop , regarding the "Create site" 'back' button, what I mean is that the wording itself was confusing to me. Perhaps this image better illustrates what I mean:



Notice that the button on the bottom has the same text "Create site", which is the final step in actually creating the site (i.e. once you press that button, the domain you have selected is provisioned and the site is created). The confusion I'm referring to is the same named button which is also the title of the previous screen "Create site". To be clear, this is with the existing flow, not any of the flows with the test flag enabled.

**Comment by chipsnyder** [comment]
Ah thanks for clarifying @mkevins I was definitely thinking you meant the change in wording. 

I'm not sure what the desired state is there was originally but I can see how that would be confusing. This would be a straightforward fix but maybe not worth doing since the goal is to release this feature soon. Although we would want to adjust it if we decide to move forward with A/B testing

**Comment by mkevins** [comment]
> This would be a straightforward fix but maybe not worth doing since the goal is to release this feature soon. Although we would want to adjust it if we decide to move forward with A/B testing

I'm not sure in what version this changed, but for version `14.0`, the "Create Site" step is earlier, and the final screen has a back button that reads "2 of 3" instead:

Back = "Create Site" | Back = "2 of 3"
-|-
IMG_0034|IMG_0033

So, it seems this is partly due to the change in ordering of the wizard steps.

**Comment by chipsnyder** [comment]
> I'm not sure in what version this changed, but for version 14.0, the "Create Site" step is earlier, and the final screen has a back button that reads "2 of 3" instead:

Thanks for bearing with me @mkevins, I think I'm still a bit lost on what's needed to help this PR be ready to merge. From what I can tell this issue is the only one blocking us from moving forward.

Let me see if I can accurately summarize to make sure I'm on the same page and we can get a game plan together to move forward:

----

Starting at `14.0` which is the reference point [mentioned](https://github.com/wordpress-mobile/WordPress-iOS/pull/15231#issuecomment-726557506) previously, we have the screens:

Title: `Create Site` <br> Back: `N/A` | Title: `1 of 3` <br> Back: `Create Site` | Title: `2 of 3` <br> Back: `1 of 3` | Title: `3 of 3` <br> Back: `2 of 3`
--- | --- | --- | ---
<kdb><a href="https://user-images.githubusercontent.com/3384451/99093484-887fbd00-25a0-11eb-8abd-b70e6f28dcc6.png"></a></kdb> | <kdb><a href="https://user-images.githubusercontent.com/3384451/99093575-a9e0a900-25a0-11eb-9923-782c8b68b8a2.png"></a></kdb> | <kdb><a href="https://user-images.githubusercontent.com/3384451/99093870-0f349a00-25a1-11eb-9246-37a077fd2ba4.png"></a></kdb> | <kdb><a href="https://user-images.githubusercontent.com/3384451/99094170-75212180-25a1-11eb-82ba-399a80dcd7de.png"></a></kdb>
[Title and back for next screen ref](https://github.com/wordpress-mobile/WordPress-iOS/blob/f0c12a5e3b9340aa7900f78506698e14e1b621f1/WordPress/Classes/ViewRelated/Site%20Creation/SiteSegments/SiteSegmentsWizardContent.swift#L61-L63) | [Title and back for next screen ref](https://github.com/wordpress-mobile/WordPress-iOS/blob/f0c12a5e3b9340aa7900f78506698e14e1b621f1/WordPress/Classes/ViewRelated/Site%20Creation/Verticals/VerticalsWizardContent.swift#L145-L147) | [Title and back for next screen ref](https://github.com/wordpress-mobile/WordPress-iOS/blob/f0c12a5e3b9340aa7900f78506698e14e1b621f1/WordPress/Classes/ViewRelated/Site%20Creation/SiteInfo/SiteInformationWizardContent.swift#L79-L81) | [Title](https://github.com/wordpress-mobile/WordPress-iOS/blob/f0c12a5e3b9340aa7900f78506698e14e1b621f1/WordPress/Classes/ViewRelated/Site%20Creation/WebAddress/WebAddressWizardContent.swift#L176-L178)


In `14.6` we [updated](https://github.com/wordpress-mobile/WordPress-iOS/pull/13679) the flow to be:

Title: `Create Site` <br> Back: `N/A` | Title: `2 of 2` <br> Back: `Create Site`
--- | --- 
 <kdb><a href="https://user-images.githubusercontent.com/3384451/99099022-d4823000-25a7-11eb-894d-b08c331988ac.png"></a></kdb> | <kdb><a href="https://user-images.githubusercontent.com/3384451/99099061-df3cc500-25a7-11eb-9274-abea0a64dbfe.png"></a></kdb>
[Title and back for next screen ref](https://github.com/wordpress-mobile/WordPress-iOS/blob/18f899f6814d8b42ab5a0ea36100625f5728f8a6/WordPress/Classes/ViewRelated/Site%20Creation/SiteSegments/SiteSegmentsWizardContent.swift#L61-L63) | [Title ref](https://github.com/wordpress-mobile/WordPress-iOS/blob/18f899f6814d8b42ab5a0ea36100625f5728f8a6/WordPress/Classes/ViewRelated/Site%20Creation/WebAddress/WebAddressWizardContent.swift#L176-L178)

Disableing the new flow in this PR we get the same results as `14.6`.

You [mentioned](https://github.com/wordpress-mobile/WordPress-iOS/pull/15231#issuecomment-726415343) the Back button on the domain screen and the call to action on that screen both say `Create Site`. I agree this could be a bit confusing if someone wanted to cancel.

---

If I'm focusing on the right issue, then my confusion is I'm not sure what the desired state, should be. Should we update this to `1 of 2` or some other descriptive name like `Choose type`? Following the pattern that was there before and removing the steps the title of this screen should be `1 of 1` and the back button title would be unchanged, which doesn't solve the issue I think we're discussing. However it seems the step `2 of 2` was [discussed](https://github.com/wordpress-mobile/WordPress-iOS/pull/13679#pullrequestreview-378640512) as being the desired output. But it doesn't look like anyone discussed the back button's title.

To provide more context the next task that I plan to start is to update the [Choose your Domain](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2722) UI which includes getting rid of the title `# of #` so whatever we go with here will be short lived.

> So, it seems this is partly due to the change in ordering of the wizard steps.

The confusion of having the same Back and Call To Action seems to have been [introduced](https://github.com/wordpress-mobile/WordPress-iOS/pull/13679) when we went down from five total steps to three. So this seems to be something that has been around for sometime. 

---

To fix the issue we would need to add: 

`navigationItem.backButtonTitle = NSLocalizedString("{Not sure what to add here}", comment: "{Descriptive info}") `

To this method:
https://github.com/wordpress-mobile/WordPress-iOS/blob/022c0ba2efdce75cac077deccaaef73091dd4f0b/WordPress/Classes/ViewRelated/Site%20Creation/SiteSegments/SiteSegmentsWizardContent.swift#L61-L63

We would then update the `{...}`'s with whatever the desired goal is. 

---
**PS.** Apologies If I'm completely lost in the woods and focusing on the wrong thing 😅 and would 💯 appreciate any course correction on the change we're looking for.

**Comment by mkevins** [comment]
Hey Chip :wave: :smile: 

At the time I encountered the screen with "Create Site" as both a back button and a confirm button, I wasn't sure when it was introduced.

> So this seems to be something that has been around for sometime.

Since you've confirmed this was already present, I see no reason to change the wording in this PR. I just didn't want to _introduce_ the issue with this PR, if that makes sense. So, it seems like all is good. :+1: 

> Although we would want to adjust it if we decide to move forward with A/B testing

I think we can address such a case if / when needed, in a later PR.

**Review by mkevins — APPROVED** [review]
Nice work Chip, LGTM!

**Review by antonis — APPROVED** [review]
Great work @chipsnyder 👍 
Both the code and the app behaviour LGTM 🎉

**Pull request merged**